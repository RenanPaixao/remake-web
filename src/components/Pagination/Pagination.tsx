import { Button, Flex } from '@chakra-ui/react'
import { FaArrowLeft, FaArrowRight } from 'react-icons/fa'

interface IProps{
  size: number
  onPrevious: () => void
  onNext: () => void
  onNumberClick: (number: number) => void
  activeNumber?: number
  gap?: number
}

export const Pagination = (props: IProps) => {
  const { size, gap,  activeNumber, onNumberClick, onPrevious, onNext } = props
  const buttonProps = {
    colorScheme: 'blue',
    variant: 'ghost'
  }
  return <Flex gap={gap}>
    {size <= 1 ? null : <Button {...buttonProps} onClick={onPrevious} isDisabled={activeNumber === 1}><FaArrowLeft /></Button>}
    {
      Array(size).fill(0).map((_, index) => {
        const current = index + 1
        return <Button
          {...buttonProps}
          isActive={current === activeNumber}
          onClick={() => onNumberClick(current)}
          key={index}>{current}
        </Button>
      })
    }
    {size <= 1 ? null : <Button {...buttonProps} onClick={onNext} isDisabled={activeNumber === size}><FaArrowRight /></Button>}
  </Flex>
}
