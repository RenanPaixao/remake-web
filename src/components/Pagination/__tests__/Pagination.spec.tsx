import { expect } from 'vitest'
import { customRender } from '../../../../tests/test-utils.tsx'
import { Pagination } from '../Pagination.tsx'
import { fireEvent } from '@testing-library/dom'
import { act } from '@testing-library/react'

const pages = 5
const onNext = vi.fn()
const onPrevious = vi.fn()
const onNumberClick = vi.fn()
describe('Questions list', () => {
  it('should render', async () => {
    const { container } = customRender(<Pagination
      size={pages}
      onNext={onNext}
      onPrevious={onPrevious}
      onNumberClick={onNumberClick}
      activeNumber={3}
    />)

    expect(container).toMatchSnapshot()
  })

  it('should render with disabled previous and next buttons case is on first or last pages', async () => {
    const { getByRole, rerender } = customRender(<Pagination
      size={pages}
      onNext={onNext}
      onPrevious={onPrevious}
      onNumberClick={onNumberClick}
      activeNumber={1}
    />)

    const nextButton = getByRole('button', { name: 'next button' })
    const previousButton = getByRole('button', { name: 'previous button' })

    expect(nextButton).toBeEnabled()
    expect(previousButton).toBeDisabled()

    rerender(<Pagination
      size={pages}
      onNext={onNext}
      onPrevious={onPrevious}
      onNumberClick={onNumberClick}
      activeNumber={pages}
    />)

    expect(nextButton).toBeDisabled()
    expect(previousButton).toBeEnabled()
  })
  it('should trigger actions', async () => {
    const { getByRole } = customRender(<Pagination
      size={pages}
      onNext={onNext}
      onPrevious={onPrevious}
      onNumberClick={onNumberClick}
      activeNumber={3}
    />)

    const nextButton = getByRole('button', { name: 'next button' })
    const previousButton = getByRole('button', { name: 'previous button' })
    const numberButton = (number: number) => {
      return getByRole('button', { name: `button page ${number}` })
    }

    await act(async () => {
      fireEvent.click(numberButton(4))
      fireEvent.click(nextButton)
      fireEvent.click(previousButton)
    })

    expect(onNext).toHaveBeenCalled()
    expect(onPrevious).toHaveBeenCalledOnce()
    expect(onNumberClick).toHaveBeenCalledOnce()
    expect(onNumberClick).toHaveBeenCalledWith(4)
  })
})
