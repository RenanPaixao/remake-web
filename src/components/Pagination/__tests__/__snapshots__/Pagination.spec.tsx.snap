// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pagination > should render 1`] = `
<div>
  <div
    class="css-k008qs"
  >
    <button
      aria-label="previous button"
      class="chakra-button css-19byqlw"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        />
      </svg>
    </button>
    <button
      aria-label="button page 1"
      class="chakra-button css-19byqlw"
      type="button"
    >
      1
    </button>
    <button
      aria-label="button page 2"
      class="chakra-button css-19byqlw"
      type="button"
    >
      2
    </button>
    <button
      aria-label="button page 3"
      class="chakra-button css-19byqlw"
      data-active=""
      type="button"
    >
      3
    </button>
    <button
      aria-label="button page 4"
      class="chakra-button css-19byqlw"
      type="button"
    >
      4
    </button>
    <button
      aria-label="button page 5"
      class="chakra-button css-19byqlw"
      type="button"
    >
      5
    </button>
    <button
      aria-label="next button"
      class="chakra-button css-19byqlw"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
        />
      </svg>
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`Questions list > should render 1`] = `
<div>
  <div
    class="css-k008qs"
  >
    <button
      aria-label="previous button"
      class="chakra-button css-19byqlw"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        />
      </svg>
    </button>
    <button
      aria-label="button page 1"
      class="chakra-button css-19byqlw"
      type="button"
    >
      1
    </button>
    <button
      aria-label="button page 2"
      class="chakra-button css-19byqlw"
      type="button"
    >
      2
    </button>
    <button
      aria-label="button page 3"
      class="chakra-button css-19byqlw"
      data-active=""
      type="button"
    >
      3
    </button>
    <button
      aria-label="button page 4"
      class="chakra-button css-19byqlw"
      type="button"
    >
      4
    </button>
    <button
      aria-label="button page 5"
      class="chakra-button css-19byqlw"
      type="button"
    >
      5
    </button>
    <button
      aria-label="next button"
      class="chakra-button css-19byqlw"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
        />
      </svg>
    </button>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
